File: Part 7 - Natural Language Processing/natural_language_processing.R
# Natural Language processing

# Importing the dataset
dataset_original = read.delim("Restaurant_Reviews.tsv", quote = '', stringsAsFactors = FALSE)

# Cleaning the text
# install.packages("tm")
# install.packages("SnowballC")
library(tm)
library(SnowballC)

corpus = VCorpus(VectorSource(dataset_original$Review))

corpus = tm_map(corpus, content_transformer(tolower)) # To lower case
corpus = tm_map(corpus, removeNumbers) # Remove numbers
corpus = tm_map(corpus, removePunctuation) # Remove punctuation
corpus = tm_map(corpus, removeWords, stopwords()) # Remove irrelevant words
corpus = tm_map(corpus, stemDocument)# Get the root of the words
corpus = tm_map(corpus, stripWhitespace)# Remove white spaces


# Creating the bag of Words model
dtm = DocumentTermMatrix(corpus)
dtm = removeSparseTerms(dtm, 0.999) # Remove infrequent words

dataset = as.data.frame(as.matrix(dtm))
dataset$Liked = dataset_original$Liked


# Encoding the target feature as factor
dataset$Liked = factor(dataset$Liked, levels = c(0, 1))

# install.packages('caTools')

# Spliting the datasetset into Training set and Test set
library(caTools)
set.seed(123)
# Split ratio is for training set in this case not for test
# The column used is y
split = sample.split(dataset$Liked, SplitRatio = 0.8)
training_set = subset(dataset, split==TRUE)
test_set = subset(dataset, split==FALSE)

# Fitting Random Forest Classification to the training set
# install.packages('randomForest')
library(randomForest)
classifier = randomForest(x = training_set[-692], y = training_set$Liked, ntree=10)

# Predicting the Test set results
y_pred = predict(classifier, type='response', newdata = test_set[-692])

# Making the confusion matrix
cm = table(test_set[,692], y_pred)

File: Part 7 - Natural Language Processing/19 -  Text classification/natural_language_processing.py
#%% Natural Language processing

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Restaurant_Reviews.tsv', delimiter = '\t', quoting = 3) # quoting = 3 is to ignore the double quotes ""

#%% Cleaning the texts
import re
import nltk
nltk.download("stopwords") # List of irrelevant words to remove

from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer # To get the roots of words

corpus = []
for i in range(len(dataset)):
    review = re.sub('[^a-zA-Z]', ' ', dataset["Review"][i]) # Remove symbols
    review = review.lower()
    review = review.split()
    ps = PorterStemmer() # Keep the root of the word
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))] # Remove irrelevant words
    review = ' '.join(review)
    corpus.append(review)


#%% Creating the Bag of Words model

from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer(max_features= 1500) # we remove the unfrequented words in the corpus
x = cv.fit_transform(corpus).toarray()
y = dataset["Liked"].values

#%% Splitting the dataset into train and test set
from sklearn.model_selection import train_test_split
# random_state is the seed
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0)

# %% Fitting the Naive Bayes to the training set
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(x_train, y_train)

# %% Predicting the test set results
y_pred = classifier.predict(x_test)
 
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)